### PR TITLE
Parallel CPU and improved GPU preconditioner

### PR DESCRIPTION
**Pull request by michel2323**
* Set default number of threads for CPU KA to `Threads.nthreads()`
* GPU preconditioner is now one kernel per step instead of `nblock` kernels
* CPU preconditioner now also uses threads

This still requires some regression testing for performance. I hope it's an improvement.

### CODE REVIEW

**Comment by michel2323** [comment]
This is also supposed to solve the issue with non-deterministic results on the GPU.

File: src/LinearSolvers/preconditioners.jl [hunk 1]
```diff
@@ -23,22 +24,45 @@ Creates an object for the block-Jacobi preconditioner
...
-    map
-    cumap
-    part
-    cupart
-    P
-    id
-    function BlockJacobiPreconditioner(J, npart, device=KA.CPU())
-        if isa(J, CUSPARSE.CuSparseMatrixCSR)
+struct BlockJacobiPreconditioner{AT,GAT,VI,GVI,MT,GMT,MI,GMI,SMT} <: AbstractPreconditioner
+    nblocks::Int64
```
Comment (frapac) [inline src/LinearSolvers/preconditioners.jl:28]: The best practice is to use `Int` instead of `Int64` (for 32 bits architecture)

File: src/LinearSolvers/preconditioners.jl [hunk 2]
```diff
@@ -55,18 +79,28 @@ mutable struct BlockJacobiPreconditioner <: AbstractPreconditioner
         for (i,v) in enumerate(part)
             push!(partitions[v], i)
         end
-        Js = Vector{Matrix{Float64}}(undef, npart)
-        nJs = maximum(length.(partitions))
-        id = Matrix{Float64}(I, nJs, nJs)
+        lpartitions = VI(undef, npart)
```
Comment (frapac) [inline src/LinearSolvers/preconditioners.jl:82]: I would replace this line with broadcasting:
`lpartitions = length.(partitions)`

File: src/LinearSolvers/preconditioners.jl [hunk 3]
```diff
@@ -55,18 +79,28 @@ mutable struct BlockJacobiPreconditioner <: AbstractPreconditioner
...
+        for i in 1:npart
+            lpartitions[i] = length(partitions[i])
+        end
+        blocksize = maximum(length.(partitions))
```
Comment (frapac) [inline src/LinearSolvers/preconditioners.jl:86]: In fact, you did the broadcasting here :-) I would rephrase this line as 
`blocksize = maximum(lpartitions)`
to avoid doing twice the same operation

File: src/LinearSolvers/preconditioners.jl [hunk 4]
```diff
@@ -55,18 +79,28 @@ mutable struct BlockJacobiPreconditioner <: AbstractPreconditioner
...
+        blocks = AT(undef, blocksize, blocksize, npart)
```
Comment (frapac) [inline src/LinearSolvers/preconditioners.jl:87]: I would be careful there. E.g. 
```julia
a = rand(100)
VT = CuVector{Float64}
@time VT(a);
  0.000097 seconds (10 allocations: 240 bytes)

@time convert(VT, a);
  0.000084 seconds (7 allocations: 192 bytes)
```
It's usually better to use `convert` rather than `VT(...)`.

Comment (frapac) [reply src/LinearSolvers/preconditioners.jl:87]: Also, if the type is already set to the destination:
```julia
VT = Vector{Float64}
@time convert(VT, a);
  0.000005 seconds
@time VT(a);
  0.000006 seconds (1 allocation: 896 bytes)
```
One version is allocating, the other not

**Review by frapac — APPROVED** [review]
Nice that now `BlockJacobiPreconditioner` is properly typed!

File: src/LinearSolvers/preconditioners.jl (hunk 3 above)
Comment (michel2323) [reply src/LinearSolvers/preconditioners.jl:86]: lol psst.

File: src/LinearSolvers/preconditioners.jl (hunk 4 above)
Comment (michel2323) [reply src/LinearSolvers/preconditioners.jl:87]: I tried to dig into the source code, because I don't understand the difference. First, if I rerun the code I get this:
```julia

julia> @time VT(a)
  0.000088 seconds (9 allocations: 224 bytes)

julia> @time convert(VT, a)
  0.000082 seconds (9 allocations: 224 bytes)
```
But that's Julia 1.5.3.

File: src/LinearSolvers/preconditioners.jl (hunk 1 above)
Comment (michel2323) [reply src/LinearSolvers/preconditioners.jl:28]: I spent too much time in code converting `int` to `long long int` due to overflows that I tend to be on the cautious side. In fact, I don't see a reason for not being explicit. The code would still work on a 32bit architecture. And if it's indeed a performance issue we should explicitly have an Integer type.

Comment (michel2323) [reply src/LinearSolvers/preconditioners.jl:28]: Which we could actually also do for the `Float64`, especially on the GPU.
